base/reduce.jl

```diff
@@ -306,25 +306,29 @@ with reduction `op` over an empty array with element type of `T`.
 
 If not defined, this will throw an `ArgumentError`.
 """
-reduce_empty(op, T) = _empty_reduce_error()
-reduce_empty(::typeof(+), T) = zero(T)
+reduce_empty(op, ::Type{T}) where T = _empty_reduce_error()
+reduce_empty(::typeof(+), ::Type{Union{}}) = _empty_reduce_error()   # avoid invalidation
```

Comment:
I think we should have this. It is not because for avoiding invalidation but for "correctness."  I think it's better to do `_empty_reduce_error()` in `sum(Union{}[])` than throwing the `MethodError`.